spec/abstractops.html

```diff
@@ -622,8 +622,8 @@ <h1>ToRelativeTemporalObject ( _options_ )</h1>
       1. If _timeZone_ is *undefined*, then
         1. Return ? CreateTemporalDate(_result_.[[Year]], _result_.[[Month]], _result_.[[Day]], _calendar_).
       1. If _offsetBehaviour_ is ~option~, then
-        1. If IsTimeZoneOffsetString(_offsetString_) is *false*, throw a *RangeError* exception.
-        1. Let _offsetNs_ be ParseTimeZoneOffsetString(_offsetString_).
+        1. If IsTimeZoneOffsetStringNanosecondPrecision(_offsetString_) is *false*, throw a *RangeError* exception.
+        1. Let _offsetNs_ be ParseTimeZoneOffsetStringNanoseconds(_offsetString_).
```

Comment:
I like the idea of a single parsing AO and a single formatting AO, and considered doing that initially. 

Where I got tripped up was what to do about IsTimeZoneOffsetString[NanosecondPrecision]. That's used in ~20 places in spec text, including a bunch inside 262 that'd have to be dragged into Temporal if we removed those AOs.

Do you think we should retain those Is* AOs?  If yes, their trivial implementation could use the (single) parsing AO.

Somewhat related: is there any way to avoid the DRY noise of the following text being repeated 15+ times?
```
1. If _offsetRecord_ is a List of errors, throw a *RangeError* exception.
  1. If _offsetRecord_.[[Precision]] is not ~hour~ or ~minute~, throw a *RangeError* exception.
```

If we're already building a custom parsing AO for offsets, could this just be one line?

What I was thinking about would be a record with these fields:
* [[Nanoseconds]] - integer or `~empty~` if not matched
* [[Minutes]] - integer or `~empty~` if not matched

I'm not sure there's a lot of value supporting other precisions (hour, second, etc.) if nothing in the spec actually uses those.

Finally, does having standalone AOs make it easier for implementers to create optimized paths for each format?  I don't have an opinion about this, just wondering.

Reply:
> Do you think we should retain those Is* AOs? If yes, their trivial implementation could use the (single) parsing AO.

Sure, as long as things are structured for their later removal.

> 
> Somewhat related: is there any way to avoid the DRY noise of the following text being repeated 15+ times?
> 
> ```
> 1. If _offsetRecord_ is a List of errors, throw a *RangeError* exception.
>   1. If _offsetRecord_.[[Precision]] is not ~hour~ or ~minute~, throw a *RangeError* exception.
> ```
> 
> If we're already building a custom parsing AO for offsets, could this just be one line?
> 
> What I was thinking about would be a record with these fields:
> 
> * [[Nanoseconds]] - integer or `~empty~` if not matched
> * [[Minutes]] - integer or `~empty~` if not matched
> 
> I'm not sure there's a lot of value supporting other precisions (hour, second, etc.) if nothing in the spec actually uses those.

I'd pretty strongly prefer to specify all numeric data in terms of the same units (i.e., nanoseconds). Having a mixture seems to invite confusion and possibly implementation bugs, and going all the way to a Duration Record is too far for the same reason you suggest here. So the shape wouldn't be `{ [[Minutes]]: integer | ~empty~, [[Nanoseconds]]: integer | ~empty~ }`, but could be e.g. `{ [[Nanoseconds]]: integer | ~empty~, [[HasSubMinutePrecision]]: Boolean }}`.

> Finally, does having standalone AOs make it easier for implementers to create optimized paths for each format? I don't have an opinion about this, just wondering.

I don't think there's a direct connection from one to the other in most cases; for me it's more of a judgment call about the cognitive burden of spec comprehension.

Reply:
> but could be e.g. `{ [[Nanoseconds]]: integer | ~empty~, [[HasSubMinutePrecision]]: Boolean }}`

👍

Reply:
I just pushed an update that changes parsing to return records. It worked better to have two different parsing methods: one for time zone identifiers and one for RFC 3339 string offsets.  Here's what I've got so far. Let me know what you think:

* Split `ParseTimeZoneOffsetString` into two new AOs for parsing ISO string offsets and offset time zone identifiers, respectively. Both return Records instead of primtives:
  * `ParseUTCOffsetString` returns { [[OffsetNanoseconds]: an integer, [[HasSubMinutePrecision]]: a Boolean }
  * `ParseTimeZoneIdentifier` returns { [[Name]]: a String or `~empty~`, [[OffsetMinutes]]: an integer or `~empty~`}
  * If the input string isn't valid, both new AOs return `~unmatched~` instead of a Record.